crates/ruff_linter/src/noqa.rs

```diff
@@ -418,182 +300,408 @@ impl<'a> FileNoqaDirectives<'a> {
+    let lexer = NoqaLexer::starts_at(offset, source);
+    lexer.lex_inline_noqa()
+}
 
-        let line = Self::lex_whitespace(line);
-        let Some(line) = Self::lex_char(line, '#') else {
-            return Ok(None);
-        };
-        let comment_start = init_line_len - line.text_len() - '#'.text_len();
-        let line = Self::lex_whitespace(line);
+/// Lexes file-level exemption comment, e.g. `# ruff: noqa: F401`
+fn lex_file_exemption(
+    comment_range: TextRange,
+    source: &str,
+) -> Result<Option<NoqaLexerOutput<'_>>, LexicalError> {
+    let source = &source[..comment_range.end().to_usize()];
+    let offset = comment_range.start();
+    let lexer = NoqaLexer::starts_at(offset, source);
+    lexer.lex_file_exemption()
+}
 
-        let Some(line) = Self::lex_flake8(line).or_else(|| Self::lex_ruff(line)) else {
-            return Ok(None);
-        };
+pub(crate) fn lex_codes(text: &str) -> Result<Vec<Code<'_>>, LexicalError> {
+    let mut lexer = NoqaLexer::starts_at(TextSize::new(0), text);
+    lexer.lex_codes()?;
+    Ok(lexer.codes)
+}
 
-        let line = Self::lex_whitespace(line);
-        let Some(line) = Self::lex_char(line, ':') else {
-            return Ok(None);
-        };
-        let line = Self::lex_whitespace(line);
-        let Some(line) = Self::lex_noqa(line) else {
-            return Ok(None);
-        };
-        let line = Self::lex_whitespace(line);
+/// Lexer for `noqa` comment lines.
+///
+/// Assumed to be initialized with range or offset starting
+/// at the `#` at the beginning of a `noqa` comment.
+#[derive(Debug)]
+struct NoqaLexer<'a> {
+    /// A slice of source text starting at the beginning of a `noqa` comment
+    line: &'a str,
+    /// Contains convenience methods for lexing
+    cursor: Cursor<'a>,
+    /// Byte offset of the start of the `noqa` comment
+    offset: TextSize,
+    /// Non-fatal warnings collected during lexing
+    warnings: Vec<LexicalWarning>,
+    /// Tracks whether we are lexing in a context with a missing delimiter
+    /// e.g. at `C` in `F401C402`.
+    missing_delimiter: bool,
+    /// Rule codes collected during lexing
+    codes: Vec<Code<'a>>,
+}
 
-        Ok(Some(if line.is_empty() {
-            // Ex) `# ruff: noqa`
-            Self::All
-        } else {
-            // Ex) `# ruff: noqa: F401, F841`
-            let Some(line) = Self::lex_char(line, ':') else {
-                return Err(ParseError::InvalidSuffix);
+impl<'a> NoqaLexer<'a> {
+    fn new(source: &'a str, range: TextRange) -> Self {
+        Self {
+            line: &source[range],
+            offset: range.start(),
+            cursor: Cursor::new(&source[range]),
+            warnings: Vec::new(),
+            missing_delimiter: false,
+            codes: Vec::new(),
+        }
+    }
+
+    /// Initialize [`NoqaLexer`] at offset.
+    ///
+    /// The offset should be the beginning of a `noqa` comment.
+    fn starts_at(offset: TextSize, source: &'a str) -> Self {
```

Comment:
I'd find it useful to document what `source` means here. Looking at the implementation, I think it's only a subset of the entire file content. 

For [`parse_string_annotation`](https://github.com/astral-sh/ruff/blob/7d8fdb04c240d5c4f9f72a0e6aae42d0739fc099/crates/ruff_python_parser/src/lib.rs#L222-L225) we consider `source` to be the entire file content and the implementation would extract out the relevant subslice.

I'm not asking you to change but it would be useful to make sure we pass the correct source here.